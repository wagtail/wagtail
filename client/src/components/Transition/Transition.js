import PropTypes from 'prop-types';
import React from 'react';

import CSSTransitionGroup from 'react-transition-group/CSSTransitionGroup';

const TRANSITION_DURATION = 210;

// The available transitions. Must match the class names in CSS.
export const PUSH = 'push';
export const POP = 'pop';

/**
 * Wrapper arround react-transition-group with default values.
 */
const Transition = ({
  name,
  component,
  className,
  duration,
  children,
}) => (
  <CSSTransitionGroup
    component={component}
    transitionEnterTimeout={duration}
    transitionLeaveTimeout={duration}
    transitionName={`c-transition-${name}`}
    className={className}
  >
    {children}
  </CSSTransitionGroup>
);

Transition.propTypes = {
  name: PropTypes.oneOf([PUSH, POP]).isRequired,
  component: PropTypes.string,
  className: PropTypes.string,
  duration: PropTypes.number,
  children: PropTypes.node,
};

Transition.defaultProps = {
  component: 'div',
  children: null,
  className: null,
  duration: TRANSITION_DURATION,
};

export default Transition;
