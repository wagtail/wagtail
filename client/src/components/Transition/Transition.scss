// =============================================================================
// Transitions
// =============================================================================

$c-transition-duration: 200ms;

.c-transition-group {
    position: absolute;
    width: 100%;
    top: 0;
}

.c-transition-push-enter {
    transform: translateX(100%);
    transition: transform $c-transition-duration ease, opacity $c-transition-duration linear;
    opacity: 0;
}

.c-transition-push-enter-active {
    transform: translateX(0);
    opacity: 1;
}

.c-transition-push-leave {
    transform: translateX(0);
    transition: transform $c-transition-duration ease, opacity $c-transition-duration linear;
    opacity: 1;
}

.c-transition-push-leave-active {
    transform: translateX(-100%);
    opacity: 0;
}

// =============================================================================
// Pop transition
// =============================================================================
.c-transition-pop-enter {
    transform: translateX(-100%);
    transition: transform $c-transition-duration ease, opacity $c-transition-duration linear;
    opacity: 0;
}

.c-transition-pop-enter-active {
    transform: translateX(0);
    opacity: 1;
}

.c-transition-pop-leave {
    transform: translateX(0);
    transition: transform $c-transition-duration ease, opacity $c-transition-duration linear;
    opacity: 1;
}

.c-transition-pop-leave-active {
    transform: translateX(100%);
    opacity: 0;
}
