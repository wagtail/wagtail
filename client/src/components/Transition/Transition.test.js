import React from 'react';
import { shallow } from 'enzyme';

import Transition, { PUSH } from './Transition';

describe('Transition', () => {
  it('exists', () => {
    expect(Transition).toBeDefined();
  });

  it('basic', () => {
    expect(shallow(<Transition name={PUSH} />)).toMatchSnapshot();
  });
});
