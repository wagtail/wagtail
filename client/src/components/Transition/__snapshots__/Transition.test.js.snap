// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transition basic 1`] = `
<CSSTransitionGroup
  className={null}
  component="div"
  transitionAppear={false}
  transitionEnter={true}
  transitionEnterTimeout={210}
  transitionLeave={true}
  transitionLeaveTimeout={210}
  transitionName="c-transition-push"
/>
`;
